### PR TITLE
`setup().extend()`

### PR DESCRIPTION
**Pull request by davidkpiano**

### CODE REVIEW

File: packages/core/src/setup.ts [hunk 1]
```diff
@@ -203,17 +163,171 @@ export function setup<
       TTag
     >
   >;
-} {
+
+  extend: <
```
[the remark this answers is not in the document]
Comment (davidkpiano) [reply packages/core/src/setup.ts:167]: Okay, that's fair

File: packages/core/test/setup.types.test.ts
Comment (Andarist) [inline packages/core/test/setup.types.test.ts:1]:
```ts
setup({})
  .extend({
    actions: {
      doOneStuff: () => {},
    },
  })
  .createMachine({
    entry: "doOtherStuff", // this should be an error
  });
```

That's why I've mentioned on Discord that before doing this change we need to make it impossible to use non-implemented things when nothing was implemented in that "group".

Comment (Andarist) [reply packages/core/test/setup.types.test.ts:1]: I have a draft for this change [here](https://github.com/statelyai/xstate/pull/4527) but I didn't have time to finish it

File: packages/core/src/setup.ts [hunk 2]
```diff
@@ -203,7 +163,134 @@ export function setup<
...
+    actions?: {
+      [K in keyof TExtActions]: ActionFunction<
+        TContext,
+        TEvent,
+        TEvent,
+        TExtActions[K],
+        ToProvidedActor<TExtChildrenMap, TExtActors>,
+        ToParameterizedObject<TExtActions>,
+        ToParameterizedObject<TExtGuards>,
+        TExtDelay
```
[the remark this answers is not in the document]
Comment (Andarist) [reply packages/core/src/setup.ts:198]: and in a similar manner, it's very likely that those `never`s should be converted to their respective "outer" type parameters:
https://github.com/statelyai/xstate/blob/39702e64d031dc7bf98fb461c249836782ca93ae/packages/core/src/setup.ts#L32
https://github.com/statelyai/xstate/blob/39702e64d031dc7bf98fb461c249836782ca93ae/packages/core/src/setup.ts#L55

the defaults for those should still be `never` - the first call in a chain doesn't have any outer type parameters after all

**Comment by davidkpiano** [comment]
Closing due to drift

**Pull request closed without merging**